### PR TITLE
Optimize LIKE for more relaxed patterns

### PR DESCRIPTION
**Pull request by xumingming**
In this PR we optimize LIKE operations for patterns which I call them
kRelaxed[Prefix|Suffix] patterns, e.g.

- kRelaxedPrefix: _a_bc%%
- kRelaxedSuffix: %%_a_bc

'Relaxed' here means there is less restrictions than their counterparts.
The algorithm of recognizing these relaxed patterns can be explained by
an example, say we have a pattern '___hello___%%', it is split into 4
sub patterns:

- [0] kSingleCharWildcard: ___
- [1] kLiteralString: hello
- [2] kSingleCharWildcard: ___
- [3] kAnyCharsWildcard: %%

Since the 'kAnyCharsWildcard' only occurs at the end of the pattern, we can
determine it is a kRelaxedPrefix pattern, and then use the first 3 fixed
sub-patterns to do the matching.

The benchmark result:

Before(kGeneric):

```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
like_generic##like_generic                                   1.34s   747.38m
----------------------------------------------------------------------------
----------------------------------------------------------------------------
like_prefix##like_prefix                                  340.30ms      2.94
like_prefix##like_relaxed_prefix_1                        334.77ms      2.99
like_prefix##like_relaxed_prefix_2                        350.70ms      2.85
like_prefix##starts_with                                    5.35ms    187.05
like_substring##like_substring                               1.26s   790.87m
like_substring##strpos                                     20.55ms     48.67
like_suffix##like_suffix                                  957.06ms      1.04
like_suffix##like_relaxed_suffix_1                        935.90ms      1.07
like_suffix##like_relaxed_suffix_2                           1.08s   926.79m
like_suffix##ends_with                                      5.35ms    187.07
```

After(kRelaxedPrefix, kRelaxedSuffix):

```
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
like_generic##like_generic                                   1.48s   674.92m
----------------------------------------------------------------------------
----------------------------------------------------------------------------
like_prefix##like_prefix                                    7.05ms    141.80
like_prefix##like_relaxed_prefix_1                          9.06ms    110.36
like_prefix##like_relaxed_prefix_2                          8.55ms    116.94
like_prefix##starts_with                                    5.34ms    187.22
like_substring##like_substring                             22.47ms     44.50
like_substring##strpos                                     20.72ms     48.27
like_suffix##like_suffix                                    7.05ms    141.82
like_suffix##like_relaxed_suffix_1                          9.08ms    110.16
like_suffix##like_relaxed_suffix_2                          8.52ms    117.30
like_suffix##ends_with                                      5.35ms    187.07
```

The speedup for kRelaxedPrefix is about 40x, speedup for kRelaxedSuffix is about 100x.

### CODE REVIEW

**Comment by xumingming** [comment]
cc @mbasmanova 

The difference between this and previous PR is the third commit([Fix kRelaxedFixed pattern](https://github.com/facebookincubator/velox/pull/8594/commits/ffe4f6b2ba99fdb371ee9768a5ed46e9f23a0389)). 

`matchRelaxedFixedForward` is a help function which returns a `std::pair`
- first: a bool indicates whether matches the pattern.
- second: an integer indicates where is cursor in the input when we finished matching if 'first' is true, -1 otherwise.

So for `kRelaxedPrefix` we just use `matchRelaxedFixedForward` is enough. For`kRelaxedFixed` we additionally check `matchResult.second == input.size()`.

File: velox/functions/lib/tests/Re2FunctionsTest.cpp [hunk 1]
```diff
@@ -640,6 +698,34 @@ TEST_F(Re2FunctionsTest, likePatternFixed) {
...
+  testLike("_ab_cde_", "_ab_cde_", true);
+  testLike("xabxcdex", "_ab_cde_", true);
+  testLike("xacxcdex", "_ab_cde_", false);
+
+  testLike("ABCDEx", "ABCDE_", true);
+  testLike("ABCDEy", "ABCDE_", true);
+  testLike("ABCDCy", "ABCDE_", false);
+  testLike("abcdey", "ABCDE_", false);
+
+  testLike("fblearner_global", "fblearner_", false);
```
Comment (mbasmanova) [inline velox/functions/lib/tests/Re2FunctionsTest.cpp:711]: These test cases use Meta-specific terms and come as a surprise. Let's use something similar to previous test cases. 

testLike("ABCDE_", "ABCDE_", true);
testLike("ABCDE__", "ABCDE_", false);
testLike("ABCDE_ZZ", "ABCDE_", false);

**Review by mbasmanova — APPROVED** [review]
Thanks.

**Review by Yuhta — APPROVED** [review]

**Comment by mbasmanova** [comment]
@xumingming We are seeing correctness issues when running tests internally. Will need to spend some time debugging. 

CC: @bikramSingh91

**Comment by xumingming** [comment]
> @xumingming We are seeing correctness issues when running tests internally. Will need to spend some time debugging.
> 
> CC: @bikramSingh91

If there are reproducible cases, I can take a look.

File: velox/functions/lib/Re2Functions.h [hunk 2]
```diff
@@ -30,33 +30,118 @@ namespace facebook::velox::functions {
...
+  /// '_pr_es_to_'.
+  kRelaxedFixed,
   /// Fixed pattern followed by one or more '%', such as 'hello%', 'foo%%%%'.
   kPrefix,
+  /// kRelaxedFixed pattern followed by one or more '%', such as '_pr_es_to_%',
+  /// '_pr_es_to_%%%%'.
+  kRelaxedPrefix,
   /// Fixed pattern preceded by one or more '%', such as '%foo', '%%%hello'.
   kSuffix,
+  /// kRelaxedFixed preceded by one or more '%', such as '%_pr_es_to_',
```
Comment (mapleFU) [inline velox/functions/lib/Re2Functions.h:48]: Does the Relaxed using the regex syntax?

**Pull request closed without merging**